### PR TITLE
add evaluate for model

### PR DESCRIPTION
**Pull request by LielinJiang**

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -991,71 +978,227 @@ def fit(
...
+                feed_list=feed_list,
+                num_workers=num_workers,
+                return_list=True)
+        else:
+            eval_loader = eval_data
+
+        self._test_dataloader = eval_loader
+        metrics_name = self._metrics_name()
+
+        steps = len(eval_loader)
```
Comment (qingqing01) [inline model.py]:
```
steps = len(eval_loader) if hasattr(eval_loader, '__len__') else None
```
If support `generator` as input in DataLoader later,  will not get the length

File: distributed.py [hunk 2]
```diff
@@ -78,9 +78,28 @@ def __iter__(self):
...
             np.random.RandomState(self.epoch).shuffle(indices)
             self.epoch += 1
+
         # subsample
-        indices = indices[self.local_rank * self.num_samples: 
-                            (self.local_rank + 1) * self.num_samples]
+        def _get_indices_by_batch_size(indices):
+            subsampled_indices = []
+            last_batch_size = self.total_size % (self.batch_size * self.nranks)
+            assert last_batch_size % self.nranks == 0
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply distributed.py]: padded before:
https://github.com/LielinJiang/hapi/blob/evaluate/distributed.py#L76

File: model.py [hunk 3]
```diff
@@ -991,71 +978,227 @@ def fit(
...
-            cbks.on_epoch_begin(epoch)
+
             # FIXME: adapt to DataLoader
             loader = train_loader
             if not isinstance(train_loader, Iterable):
                 loader = train_loader()
-            logs = _run_one_epoch(loader, cbks, 'train')
+            logs = self._run_one_epoch(
+                loader, cbks, 'train', metrics_name, epoch=epoch)
             cbks.on_epoch_end(epoch, logs)
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply model.py]: done, thanks

File: model.py [hunk 4]
```diff
@@ -991,71 +978,227 @@ def fit(
...
 
             if do_eval and epoch % eval_freq == 0:
-                cbks.on_begin('eval', logs)
                 # FIXME: adapt to DataLoader
                 loader = eval_loader
                 if not isinstance(eval_loader, Iterable):
                     loader = eval_loader()
-                logs = _run_one_epoch(loader, cbks, 'eval')
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply model.py]: done, thanks

File: model.py [hunk 5]
```diff
@@ -991,71 +978,227 @@ def fit(
...
+        FIXME: add more comments and usage
+        Args:
+            eval_data (Dataset|DataLoader): An iterable data loader is used for
+                evaluation. An instance of paddle.fluid.io.Dataset or 
+                paddle.fluid.io.Dataloader is recomended.
+            batch_size (int): Integer number. The batch size of train_data and eval_data. 
+                When train_data and eval_data are both the instance of Dataloader, this 
+                parameter will be ignored.
+            log_freq (int): The frequency, in number of steps, the training logs
+                is printed.
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply model.py]: done ,thanks

File: model.py [hunk 6]
```diff
@@ -991,71 +978,227 @@ def fit(
...
+            steps=steps,
+            log_freq=log_freq,
+            verbose=verbose,
+            metrics=self._metrics_name(), )
+
+        loader = eval_loader
+        if not isinstance(eval_loader, Iterable):
+            loader = eval_loader()
+        logs = self._run_one_epoch(loader, cbks, 'eval', metrics_name)
+        cbks.on_end('eval', logs)
```
[the remark this answers is not in the document]
Comment (LielinJiang) [reply model.py]: done, thanks

File: model.py (hunk 1 above)
Comment (LielinJiang) [reply model.py]: done, thanks

**Pull request merged**